xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -359,16 +333,20 @@ public Result selectConfig(PickSubchannelArgs args) {
       }
       String cluster = null;
       Route selectedRoute = null;
-      HttpFault selectedFaultConfig;
+      RoutingConfig savedRoutingConfig;
+      Map<String, FilterConfig> selectedOverrideConfigs;
+      List<ClientInterceptor> filterInterceptors = new ArrayList<>();
       do {
-        selectedFaultConfig = routingConfig.faultConfig;
-        for (Route route : routingConfig.routes) {
+        savedRoutingConfig = routingConfig;
```

Comment:
As I was about to reply on another PR. You should _not_ read volatile the in each iteration of the loop. That has no difference than just directly using the volatile. The purpose of saving a local copy of the volatile (as what I was doing in the other PR) is to ensure the data used while executing this method (AKA, `selectConfig`) is _consistent_, even if it may become out-of-date in the middle of the execution. That is, we make the read of the volatile atomic, which is at the beginning of executing this method.

Reply:
No, must always read the latest `routingConfig` in each iteration of the loop so that you can successfully retain a cluster. A cluster in an obsolete `routingConfig` will never be retained if the cluster is removed by a racing update https://github.com/grpc/grpc-java/blob/v1.36.0/xds/src/main/java/io/grpc/xds/XdsNameResolver.java#L650. The while loop may then never make progress.